// @zoltu/ethereum-fetch-json-rpc/output-node/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ethereum_types_1 = require("@zoltu/ethereum-types");
const index_1 = require("./vendor/rlp-encoder/index");
const error_with_data_1 = require("./error-with-data");
exports.ErrorWithData = error_with_data_1.ErrorWithData;
const sleep_1 = require("./sleep");
const utils_1 = require("./utils");
class FetchJsonRpc {
    constructor(jsonRpcEndpoint, fetch, options) {
        this.jsonRpcEndpoint = jsonRpcEndpoint;
        this.fetch = fetch;
        this.sendEth = async (destination, amount) => await this.executeTransaction({ to: destination, value: amount });
        this.deployContract = async (bytecode, value) => (await this.executeTransaction({ to: null, data: bytecode, value: value })).contractAddress;
        this.onChainContractCall = async (transaction) => this.executeTransaction(transaction);
        this.offChainContractCall = async (transaction) => {
            const offChainTransaction = {
                from: transaction.from !== undefined ? transaction.from : await this.coinbase().catch(() => null) || 0n,
                to: transaction.to,
                value: transaction.value || 0n,
                data: transaction.data || new ethereum_types_1.Bytes(),
                gasLimit: transaction.gasLimit || 1000000000n,
                gasPrice: transaction.gasPrice || await this.getGasPrice(),
            };
            return await this.call(offChainTransaction);
        };
        this.executeTransaction = async (transaction) => {
            const transactionHash = await this.submitTransaction(transaction);
            let receipt = await this.getTransactionReceipt(transactionHash);
            // TODO: find out if Parity, Geth or MM return a receipt with a null block anymore (docs suggest no)
            while (receipt === null || receipt.blockNumber === null) {
                await sleep_1.sleep(1000);
                receipt = await this.getTransactionReceipt(transactionHash);
            }
            if (!receipt.status)
                throw new Error(`Transaction mined, but failed.`);
            if (!receipt.contractAddress && !transaction.to)
                throw new Error(`Contract deployment failed.  Contract address was null.`);
            return receipt;
        };
        this.submitTransaction = async (transaction) => {
            const { encodedTransaction } = await this.signTransaction(transaction);
            return await this.sendRawTransaction(encodedTransaction);
        };
        this.makeRequest = (Request, Response) => async (...args) => {
            const request = new Request(0, ...args);
            const rawRequest = request.wireEncode();
            const rawResponse = await this.remoteProcedureCall(rawRequest);
            const response = new Response(rawResponse);
            return response.result;
        };
        this.call = this.makeRequest(ethereum_types_1.Rpc.Eth.Call.Request, ethereum_types_1.Rpc.Eth.Call.Response);
        this.estimateGas = this.makeRequest(ethereum_types_1.Rpc.Eth.EstimateGas.Request, ethereum_types_1.Rpc.Eth.EstimateGas.Response);
        this.getBalance = this.makeRequest(ethereum_types_1.Rpc.Eth.GetBalance.Request, ethereum_types_1.Rpc.Eth.GetBalance.Response);
        this.getBlockByHash = this.makeRequest(ethereum_types_1.Rpc.Eth.GetBlockByHash.Request, ethereum_types_1.Rpc.Eth.GetBlockByHash.Response);
        this.getBlockByNumber = this.makeRequest(ethereum_types_1.Rpc.Eth.GetBlockByNumber.Request, ethereum_types_1.Rpc.Eth.GetBlockByNumber.Response);
        this.getBlockNumber = this.makeRequest(ethereum_types_1.Rpc.Eth.BlockNumber.Request, ethereum_types_1.Rpc.Eth.BlockNumber.Response);
        this.getBlockTransactionCountByHash = this.makeRequest(ethereum_types_1.Rpc.Eth.GetBlockTransactionCountByHash.Request, ethereum_types_1.Rpc.Eth.GetBlockTransactionCountByHash.Response);
        this.getBlockTransactionCountByNumber = this.makeRequest(ethereum_types_1.Rpc.Eth.GetBlockTransactionCountByNumber.Request, ethereum_types_1.Rpc.Eth.GetBlockTransactionCountByNumber.Response);
        this.getChainId = this.makeRequest(ethereum_types_1.Rpc.Eth.ChainId.Request, ethereum_types_1.Rpc.Eth.ChainId.Response);
        this.getCode = this.makeRequest(ethereum_types_1.Rpc.Eth.GetCode.Request, ethereum_types_1.Rpc.Eth.GetCode.Response);
        this.getLogs = this.makeRequest(ethereum_types_1.Rpc.Eth.GetLogs.Request, ethereum_types_1.Rpc.Eth.GetLogs.Response);
        this.getProof = this.makeRequest(ethereum_types_1.Rpc.Eth.GetProof.Request, ethereum_types_1.Rpc.Eth.GetProof.Response);
        this.getStorageAt = this.makeRequest(ethereum_types_1.Rpc.Eth.GetStorageAt.Request, ethereum_types_1.Rpc.Eth.GetStorageAt.Response);
        this.getTransactionByBlockHashAndIndex = this.makeRequest(ethereum_types_1.Rpc.Eth.GetTransactionByBlockHashAndIndex.Request, ethereum_types_1.Rpc.Eth.GetTransactionByBlockHashAndIndex.Response);
        this.getTransactionByBlockNumberAndIndex = this.makeRequest(ethereum_types_1.Rpc.Eth.GetTransactionByBlockNumberAndIndex.Request, ethereum_types_1.Rpc.Eth.GetTransactionByBlockNumberAndIndex.Response);
        this.getTransactionByHash = this.makeRequest(ethereum_types_1.Rpc.Eth.GetTransactionByHash.Request, ethereum_types_1.Rpc.Eth.GetTransactionByHash.Response);
        this.getTransactionCount = this.makeRequest(ethereum_types_1.Rpc.Eth.GetTransactionCount.Request, ethereum_types_1.Rpc.Eth.GetTransactionCount.Response);
        // workaround for Parity returning partial transaction receipts before mining
        // public readonly getTransactionReceipt = this.makeRequest(Rpc.Eth.GetTransactionReceipt.Request, Rpc.Eth.GetTransactionReceipt.Response)
        this.getTransactionReceipt = async (transactionHash) => {
            const request = new ethereum_types_1.Rpc.Eth.GetTransactionReceipt.Request(null, transactionHash);
            const rawRequest = request.wireEncode();
            const rawResponse = await this.remoteProcedureCall(rawRequest);
            if (rawResponse.result === null || rawResponse.result.blockNumber === null || rawResponse.result.blockHash === null)
                return null;
            const response = new ethereum_types_1.Rpc.Eth.GetTransactionReceipt.Response(rawResponse);
            return response.result;
        };
        this.getUncleByBlockHashAndIndex = this.makeRequest(ethereum_types_1.Rpc.Eth.GetUncleByBlockHashAndIndex.Request, ethereum_types_1.Rpc.Eth.GetUncleByBlockHashAndIndex.Response);
        this.getUncleByBlockNumberAndIndex = this.makeRequest(ethereum_types_1.Rpc.Eth.GetUncleByBlockNumberAndIndex.Request, ethereum_types_1.Rpc.Eth.GetUncleByBlockNumberAndIndex.Response);
        this.getUncleCountByBlockHash = this.makeRequest(ethereum_types_1.Rpc.Eth.GetUncleCountByBlockHash.Request, ethereum_types_1.Rpc.Eth.GetUncleCountByBlockHash.Response);
        this.getUncleCountByBlockNumber = this.makeRequest(ethereum_types_1.Rpc.Eth.GetUncleCountByBlockNumber.Request, ethereum_types_1.Rpc.Eth.GetUncleCountByBlockNumber.Response);
        this.getProtocolVersion = this.makeRequest(ethereum_types_1.Rpc.Eth.ProtocolVersion.Request, ethereum_types_1.Rpc.Eth.ProtocolVersion.Response);
        this.sendRawTransaction = this.makeRequest(ethereum_types_1.Rpc.Eth.SendRawTransaction.Request, ethereum_types_1.Rpc.Eth.SendRawTransaction.Response);
        this.sendTransaction = this.submitTransaction;
        this.signTransaction = async (transaction) => {
            const gasEstimatingTransaction = {
                from: transaction.from !== undefined ? transaction.from : await this.coinbase().catch(() => null) || 0n,
                to: transaction.to,
                value: transaction.value || 0n,
                data: transaction.data || new ethereum_types_1.Bytes(),
                gasLimit: transaction.gasLimit || 1000000000n,
                gasPrice: transaction.gasPrice || await this.getGasPrice(),
            };
            const unsignedTransaction = {
                ...gasEstimatingTransaction,
                gasLimit: transaction.gasLimit || await this.gasLimitProvider(gasEstimatingTransaction, this.estimateGas),
                nonce: transaction.nonce || await this.getTransactionCount(gasEstimatingTransaction.from, 'pending'),
                chainId: await this.chainId,
            };
            if (this.signatureProvider === undefined) {
                return await this.makeRequest(ethereum_types_1.Rpc.Eth.SignTransaction.Request, ethereum_types_1.Rpc.Eth.SignTransaction.Response)(unsignedTransaction);
            }
            else {
                const rlpEncodedUnsignedTransaction = this.rlpEncodeTransaction(unsignedTransaction);
                const signature = await this.signatureProvider(rlpEncodedUnsignedTransaction);
                const v = (signature.yParity === 'even' ? 0n : 1n) + 35n + 2n * unsignedTransaction.chainId;
                const decodedTransaction = { ...unsignedTransaction, r: signature.r, s: signature.s, v };
                const encodedTransaction = this.rlpEncodeTransaction(decodedTransaction);
                return { decodedTransaction, encodedTransaction };
            }
        };
        this.sign = async (signerAddress, data) => {
            if (this.signatureProvider === undefined)
                return this.makeRequest(ethereum_types_1.Rpc.Eth.Sign.Request, ethereum_types_1.Rpc.Eth.Sign.Response)(signerAddress, data);
            if (await this.coinbase() !== signerAddress)
                throw new Error(`Cannot sign messages for address 0x${signerAddress.toString(16).padStart(40, '0')}`);
            const messageToSign = this.mutateMessageForSigning(data);
            const signature = await this.signatureProvider(messageToSign);
            return this.encodeSignature(signature);
        };
        this.syncing = this.makeRequest(ethereum_types_1.Rpc.Eth.Syncing.Request, ethereum_types_1.Rpc.Eth.Syncing.Response);
        this.remoteProcedureCall = async (request) => {
            const requestBodyJson = JSON.stringify(request);
            const response = await this.fetch(this.jsonRpcEndpoint, { method: 'POST', body: requestBodyJson, headers: { 'Content-Type': 'application/json' } });
            if (!response.ok)
                throw new error_with_data_1.ErrorWithData(`${response.status}: ${response.statusText}\n${response.text()}`, request);
            const responseBody = await response.json();
            ethereum_types_1.validateJsonRpcResponse(responseBody);
            if (ethereum_types_1.isJsonRpcError(responseBody))
                throw new error_with_data_1.ErrorWithData(this.extractErrorMessage(responseBody.error), { request, code: responseBody.error.code, data: responseBody.error.data });
            return responseBody;
        };
        this.rlpEncodeTransaction = (transaction) => {
            const toEncode = [
                utils_1.stripLeadingZeros(ethereum_types_1.Bytes.fromUnsignedInteger(transaction.nonce, 256)),
                utils_1.stripLeadingZeros(ethereum_types_1.Bytes.fromUnsignedInteger(transaction.gasPrice, 256)),
                utils_1.stripLeadingZeros(ethereum_types_1.Bytes.fromUnsignedInteger(transaction.gasLimit, 256)),
                utils_1.stripLeadingZeros(transaction.to !== null ? ethereum_types_1.Bytes.fromUnsignedInteger(transaction.to, 256) : new Uint8Array(0)),
                utils_1.stripLeadingZeros(ethereum_types_1.Bytes.fromUnsignedInteger(transaction.value, 256)),
                new Uint8Array(transaction.data),
            ];
            if (!this.isSignedTransaction(transaction)) {
                toEncode.push(utils_1.stripLeadingZeros(ethereum_types_1.Bytes.fromUnsignedInteger(transaction.chainId, 256)));
                toEncode.push(utils_1.stripLeadingZeros(new Uint8Array(0)));
                toEncode.push(utils_1.stripLeadingZeros(new Uint8Array(0)));
            }
            else {
                toEncode.push(utils_1.stripLeadingZeros(ethereum_types_1.Bytes.fromUnsignedInteger(transaction.v, 256)));
                toEncode.push(utils_1.stripLeadingZeros(ethereum_types_1.Bytes.fromUnsignedInteger(transaction.r, 256)));
                toEncode.push(utils_1.stripLeadingZeros(ethereum_types_1.Bytes.fromUnsignedInteger(transaction.s, 256)));
            }
            return ethereum_types_1.Bytes.fromByteArray(index_1.rlpEncode(toEncode));
        };
        this.isSignedTransaction = (transaction) => transaction.r !== undefined;
        this.mutateMessageForSigning = (message) => {
            message = (typeof message === 'string') ? new TextEncoder().encode(message) : message;
            const messagePrefix = new TextEncoder().encode(`\x19Ethereum Signed Message:\n${message.length.toString(10)}`);
            return ethereum_types_1.Bytes.fromByteArray([...messagePrefix, ...message]);
        };
        this.encodeSignature = (signature) => {
            const v = (signature.yParity === 'even' ? 0n : 1n) + 27n;
            const rSegment = ethereum_types_1.Bytes.fromUnsignedInteger(signature.r, 256);
            const sSegment = ethereum_types_1.Bytes.fromUnsignedInteger(signature.s, 256);
            const vSegment = ethereum_types_1.Bytes.fromUnsignedInteger(v, 8);
            return ethereum_types_1.Bytes.fromByteArray([...rSegment, ...sSegment, ...vSegment]);
        };
        this.extractErrorMessage = (error) => {
            if (typeof error.data !== 'string')
                return error.message;
            // handle contract revert errors from Parity
            if (error.data.startsWith('Reverted 0x08c379a0')) {
                const offset = Number.parseInt(error.data.substr('Reverted 0x08c379a0'.length, 64), 16) * 2;
                const length = Number.parseInt(error.data.substr('Reverted 0x08c379a0'.length + offset, 64), 16) * 2;
                const message = new TextDecoder().decode(ethereum_types_1.Bytes.fromHexString(error.data.substr(19 + offset + 64, length)));
                return `Contract Error: ${message}`;
            }
            // handle contract revert errors from Nethermind
            if (error.data.startsWith('revert: ')) {
                const message = error.data.substr('revert: '.length);
                return `Contract Error: ${message}`;
            }
            return error.message;
        };
        this.gasPriceInAttoethProvider = options.gasPriceInAttoethProvider;
        this.gasLimitProvider = options.gasLimitProvider || (async (transaction, estimator) => await estimator(transaction));
        this.addressProvider = options.addressProvider;
        this.signatureProvider = options.signatureProvider;
        this.chainId = (options.chainId !== undefined) ? Promise.resolve(options.chainId) : this.getChainId();
        this.coinbase = (this.addressProvider) ? this.addressProvider : this.makeRequest(ethereum_types_1.Rpc.Eth.Coinbase.Request, ethereum_types_1.Rpc.Eth.Coinbase.Response);
        // necessary to capture value for async call, since readonly modifier doesn't take effect until after constructor
        const getSignerAddress = this.addressProvider;
        this.getAccounts = (getSignerAddress !== undefined) ? async () => [await getSignerAddress()] : this.makeRequest(ethereum_types_1.Rpc.Eth.Accounts.Request, ethereum_types_1.Rpc.Eth.Accounts.Response);
        this.getGasPrice = (this.gasPriceInAttoethProvider) ? this.gasPriceInAttoethProvider : this.makeRequest(ethereum_types_1.Rpc.Eth.GasPrice.Request, ethereum_types_1.Rpc.Eth.GasPrice.Response);
        // silence NodeJS/Chrome warnings about unhandled rejections when we pre-fetch
        this.chainId.catch(() => { });
    }
}
exports.FetchJsonRpc = FetchJsonRpc;
